hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/EmbeddableMappingTypeImpl.java

```diff
@@ -693,6 +814,60 @@ public <T> DomainResult<T> createDomainResult(
 		);
 	}
 
+	@Override
+	public boolean declaresAttribute(String embeddableClassName, AttributeMapping attributeMapping) {
+		if ( declaredAttributesBySubclass == null ) {
+			return true;
+		}
+		final Set<AttributeMapping> declaredAttributes = declaredAttributesBySubclass.get( embeddableClassName );
+		return declaredAttributes != null && declaredAttributes.contains( attributeMapping );
+	}
+
+	@Override
+	public Object getValue(Object instance, int position) {
+		final AttributeMapping attributeMapping = getAttributeMapping( position );
+		if ( declaresAttribute( instance.getClass().getName(), attributeMapping ) ) {
+			return attributeMapping.getValue( instance );
+		}
+		return null;
+	}
+
+	@Override
+	protected Object[] getAttributeValues(Object compositeInstance) {
+		final Object[] results = new Object[getNumberOfAttributeMappings()];
+		final String compositeClassName = compositeInstance.getClass().getName();
+		for ( int i = 0; i < getNumberOfAttributeMappings(); i++ ) {
+			final AttributeMapping attributeMapping = getAttributeMapping( i );
+			if ( declaresAttribute( compositeClassName, attributeMapping ) ) {
+				final Getter getter = attributeMapping.getAttributeMetadata()
+						.getPropertyAccess()
+						.getGetter();
+				results[i] = getter.get( compositeInstance );
+			}
+			else {
+				results[i] = null;
+			}
+		}
+		return results;
+	}
+
+	@Override
+	protected void setAttributeValues(Object component, Object[] values) {
+		final String compositeClassName = component.getClass().getName();
```

Comment:
Maybe it's worth delegating to `super.setAttributeValues()` if the type is not polymorphic?